### PR TITLE
[core] Render pipeline refactoring 

### PR DESCRIPTION
**Pull request by pozdnyakov**
This patch does the following:
* provides generic mechanism to synchronize buckets and layer paint properties (applied for the layers that render geometry tiles data, i.e. `symbol`, `line`, `fill`, `fill-extrusion`, `heatmap`, `circle`)
* introduces new immutable `style::LayerProperties` and `LayerRenderData` types, which is a step towards separate render thread.
* encapsulates layer paint properties initialization
* encapsulates render tile initialization
* multiple simplifications and refactoring

Fixes https://github.com/mapbox/mapbox-gl-native/issues/13812

### CODE REVIEW

File: src/mbgl/layermanager/fill_layer_factory.cpp [hunk 1]
```diff
@@ -27,8 +27,10 @@ std::unique_ptr<style::Layer> FillLayerFactory::createLayer(const std::string& i
 std::unique_ptr<Layout>
 FillLayerFactory::createLayout(const LayoutParameters& parameters,
                                std::unique_ptr<GeometryTileLayer> layer,
-                               const std::vector<const RenderLayer*>& group) noexcept {
-    return std::make_unique<PatternLayout<FillBucket>>(parameters.bucketParameters, group, std::move(layer), parameters.imageDependencies);
+                               const std::vector<Immutable<style::LayerPropertiesWrapper>>& group) noexcept {
+    using namespace style;
+    using LayoutType = PatternLayout<FillBucket, FillPropertiesWrapper, FillPattern, FillPaintProperties::PossiblyEvaluated>;
```
Comment (alexshalamov) [inline src/mbgl/layermanager/fill_layer_factory.cpp]: `*PropertiesWrapper` already knows about `*PaintProperties::PossiblyEvaluated`, we can infer type from wrapper, will remove one template parameter

File: src/mbgl/style/layers/background_layer_properties.hpp [hunk 2]
```diff
@@ -30,5 +32,20 @@ class BackgroundPaintProperties : public Properties<
     BackgroundOpacity
 > {};
 
+class BackgroundPropertiesWrapper final : public LayerPropertiesWrapper {
```
Comment (alexshalamov) [inline src/mbgl/style/layers/background_layer_properties.hpp]: Technically, looks good, except the name :smile: It is not a simple wrapper anymore, as it contains original style layer and few other things.

File: src/mbgl/renderer/tile_pyramid.cpp [hunk 3]
```diff
@@ -238,6 +238,27 @@ void TilePyramid::update(const std::vector<Immutable<style::LayerPropertiesWrapp
...
+        if (!tile.isRenderable()) continue;
+
+        const bool holdForFade = tile.holdForFade();
+        for (const auto& layerProperties : layers) {
+            const auto* typeInfo = layerProperties->baseImpl->getTypeInfo();
+            if (holdForFade && typeInfo->fadingTiles == LayerTypeInfo::FadingTiles::NotRequired) {
+                continue;
+            }
+            // Update layer properties for complete tiles; for incomplete just check the presence.
+            bool layerIsInTile = tile.isComplete() ? tile.updateLayerProperties(layerProperties)
```
Comment (alexshalamov) [inline src/mbgl/renderer/tile_pyramid.cpp]: Maybe rename `layerIsInTile` to `layerIsRenderable`?

File: src/mbgl/style/layers/background_layer_properties.hpp (hunk 2 above)
Comment (pozdnyakov) [reply src/mbgl/style/layers/background_layer_properties.hpp]: Yeah, I've been looking for a good one and did not find yet.. any options? From OOP point of view it is "Memento", however `LayerPropertiesMemento` does not look good :smile:

File: src/mbgl/renderer/tile_pyramid.cpp (hunk 3 above)
Comment (pozdnyakov) [reply src/mbgl/renderer/tile_pyramid.cpp]: `renderableByTile`

File: src/mbgl/style/layers/background_layer_properties.hpp (hunk 2 above)
Comment (pozdnyakov) [reply src/mbgl/style/layers/background_layer_properties.hpp]: Let's call it just `LayerProperties` and e.g. `CircleLayerProperties`..

File: src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp [hunk 4]
```diff
@@ -78,22 +76,27 @@ void RenderFillExtrusionLayer::render(PaintParameters& parameters, RenderSource*
             "fill extrusion",
             { *renderTexture, Color{ 0.0f, 0.0f, 0.0f, 0.0f }, depthClearValue, {} });
 
-        auto draw = [&](auto& programInstance, const auto& tileBucket, auto&& uniformValues,
+        auto draw = [&](auto& programInstance,
+                        const auto& evaluated_,
+                        const auto& crossfade_,
+                        const auto& tileBucket,
+                        auto&& uniformValues,
```
Comment (alexshalamov) [inline src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp]: const auto&

File: src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp [hunk 5]
```diff
@@ -78,22 +76,27 @@ void RenderFillExtrusionLayer::render(PaintParameters& parameters, RenderSource*
...
                         const optional<ImagePosition>& patternPositionA,
-                        const optional<ImagePosition>& patternPositionB, auto&& textureBindings) {
+                        const optional<ImagePosition>& patternPositionB,
+                        auto&& textureBindings) {
```
Comment (alexshalamov) [inline src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp]: const auto&

and remove moves in draw call below

File: src/mbgl/tile/geometry_tile_worker.cpp [hunk 6]
```diff
@@ -412,7 +392,7 @@ void GeometryTileWorker::parse() {
             }
 
             for (const auto& layer : group) {
-                buckets.emplace(layer->baseImpl->id, bucket);
+                renderData.emplace(layer->baseImpl->id, LayerRenderData{bucket, layer});
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply src/mbgl/tile/geometry_tile_worker.cpp]: bucket is shared between several items

File: src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp (hunk 4 above)
Comment (pozdnyakov) [reply src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp]: it is invoked with a temporary

**Pull request merged**